src/thunderbots/software/geom/rect.h

```diff
@@ -0,0 +1,215 @@
+     *
+     * @return the south-east corner of the rectangle.
+     */
+    constexpr Point seCorner() const;
+
+    /**
+     * Returns 0-sw 1-nw 2-ne 3-se corner for pos%4
+     *
+     * @param pos of corner wanted
+     *
+     * @return Point corresponding to position
+     */
+    constexpr Point operator[](unsigned int pos) const;
+
+    /**
+     * Translates the rectangle.
+     *
+     * @param offset the distance to move the rectangle.
+     */
+    void translate(const Point &offset);
+
+    /**
+     * Checks whether a point is within the boundries of the rectangle
+     *
+     * @param p the point to test
+     *
+     * @return bool whether the point is inside the boundry of the rectangle
+     */
+    constexpr bool containsPoint(Point p) const;
+
+    /**
+     * Tries to move all of the edges of the rectangle outwards/inwards towards
+     * the centre by "amount"
+     * while keeping the location of the centre of the rectangle the same the
+     * rectangle
+     * will not shrink to something smaller than a point
+     *
+     * @param amount the amount to shrink the recatngle by on all sides
+     * (positive or negative numbers ok)
+     *
+     * @return bool whether it was possible to expand/shrink the rectangle by
+     * amount requested
+     */
+    bool expand(double amount);
+
+    /**
+     * Gives the distance between a point and the nearest point on the rectangle
+     * boundry
+     *
+     * @param p the point to test
+     *
+     * @return double
+     */
+    double distToBoundary(Point p);
+
+   private:
+    Point min_corner;
+    Point diag;
+};
+
+inline constexpr Rect::Rect(const Point &point1, const Point &point2)
+    : min_corner(
+          point1.x() < point2.x() ? point1.x() : point2.x(),
+          point1.y() < point2.y() ? point1.y() : point2.y()),
+      diag(
+          point1.x() < point2.x() ? point2.x() - point1.x() : point1.x() - point2.x(),
+          point1.y() < point2.y() ? point2.y() - point1.y() : point1.y() - point2.y())
+{
+}
+
+inline constexpr Rect::Rect(const Point &sw_corner, double width, double height)
+    : min_corner(sw_corner), diag(width, height)
+{
+}
+
+inline constexpr double Rect::width() const
+{
+    return diag.x();
+}
+
+inline constexpr double Rect::height() const
+{
+    return diag.y();
+}
+
+inline constexpr double Rect::area() const
+{
+    return diag.x() * diag.y();
+}
+
+inline constexpr Point Rect::centre() const
+{
+    return min_corner + diag / 2;
+}
+
+inline constexpr Point Rect::neCorner() const
+{
+    return min_corner + diag;
+}
+
+inline constexpr Point Rect::nwCorner() const
+{
+    return min_corner + Point(0, diag.y());
+}
+
+inline constexpr Point Rect::swCorner() const
+{
+    return min_corner;
+}
+
+inline constexpr Point Rect::seCorner() const
+{
+    return min_corner + Point(diag.x(), 0);
+}
+
+inline constexpr Point Rect::operator[](unsigned int pos) const
+{
+    return (pos % 4) == 0
```

Comment:
This should be refactored with a switch statement on {0, 1, 2, 3} (I think?). I have no idea what usage of this would use any index > 3.
Exception should be thrown if out-of-index.

Reply:
*out of range

Reply:
yes